// react-label/webpack.config.js
const webpack = require('webpack');
const path = require('path');
const postcss = require('postcss');

const sourcePath = path.join(__dirname, './src');

module.exports = {
  context: sourcePath,
  entry: {
    'react-label': './react-label',
  },
  output: {
    libraryTarget: 'umd',
    library: 'ReactLabel',
    path: path.join(__dirname, 'dist'),
    filename: '[name].js',
  },
  resolve: {
    modules: ['node_modules', sourcePath],
    extensions: ['.js', '.jsx'],
  },
  externals: [
    {
      'react-dom': {
        root: 'ReactDOM',
        commonjs2: 'react-dom',
        commonjs: 'react-dom',
        amd: 'react-dom'
      }
    },
    {
      'react': {
        root: 'React',
        commonjs2: 'react',
        commonjs: 'react',
        amd: 'react'
      }
    },
  ],
  module: {
    rules: [{
      test: /\.jsx$/,
      exclude: /node_modules/,
      use: ['babel-loader'],
    }, {
      test: /\.css$/,
      exclude: /node_modules/,
      use: [
        'style-loader',
        'css-loader',
        {
          loader: 'postcss-loader',
          options: {
            plugins: [
              require('postcss-import')({ addDependencyTo: webpack }),
              require('postcss-cssnext'),
              require('postcss-custom-properties'),
              require('postcss-css-variables'),
              require('precss'),
              require('postcss-calc'),
              require('postcss-color-function'),
            ],
          },
        },
      ],
    }],
  },
};
